src/tensors/abstracttensor.jl

```diff
@@ -226,6 +226,55 @@ Return the dimensions of the block of a tensor corresponding to a coupled sector
 Return an iterator over all splitting - fusion tree pairs of a tensor.
 """ fusiontrees(::AbstractTensorMap)
 
+# Similar
+#---------
+# The implementation is written for similar(t, TorA, V::TensorMapSpace) -> TensorMap
+# and all other methods are just filling in default arguments
+# 4 arguments
+@doc """
+    similar(t::AbstractTensorMap, [AorT=storagetype(t)], [V=space(t)])
+    similar(t::AbstractTensorMap, [AorT=storagetype(t)], codomain, domain)
+
+Creates an uninitialized mutable tensor with the given scalar or storagetype `AorT` and
+structure `V` or `codomain ← domain`, based on the source tensormap. The second and third
+arguments are both optional, defaulting to the given tensor's `storagetype` and `space`.
+The structure may be specified either as a single `HomSpace` argument or as `codomain` and
+`domain`.
+
+By default, this will result in `TensorMap{T}(undef, V)` when custom objects do not
+specialize this method.
+""" Base.similar(::AbstractTensorMap, args...)
+
+function Base.similar(t::AbstractTensorMap, ::Type{T}, codomain::TensorSpace{S},
+                      domain::TensorSpace{S}) where {T,S}
+    return similar(t, T, codomain ← domain)
+end
+# 3 arguments
+function Base.similar(t::AbstractTensorMap, codomain::TensorSpace{S},
+                      domain::TensorSpace{S}) where {S}
+    return similar(t, storagetype(t), codomain ← domain)
+end
+function Base.similar(t::AbstractTensorMap, ::Type{T}, codomain::TensorSpace) where {T}
+    return similar(t, T, codomain ← one(codomain))
+end
+# 2 arguments
+function Base.similar(t::AbstractTensorMap, codomain::TensorSpace)
+    return similar(t, storagetype(t), codomain ← one(codomain))
+end
+Base.similar(t::AbstractTensorMap, P::TensorMapSpace) = similar(t, storagetype(t), P)
+Base.similar(t::AbstractTensorMap, ::Type{T}) where {T} = similar(t, T, space(t))
+# 1 argument
+Base.similar(t::AbstractTensorMap) = similar(t, storagetype(t), space(t))
+
+# generic implementation for AbstractTensorMap -> returns `TensorMap`
+function Base.similar(::AbstractTensorMap, ::Type{TorA},
+                      P::TensorMapSpace{S}) where {TorA<:MatOrNumber,S}
+    N₁ = length(codomain(P))
+    N₂ = length(domain(P))
+    TT = tensormaptype(S, N₁, N₂, TorA)
+    return TT(undef, codomain(P), domain(P))
```

Comment:
`TT(undef, P)` ? I guess it depends on which constructor is "the most fundamental"

Reply:
the problem here is that this is a fully specified tensormap type, and I did not add a constructor for this. The "fundamental" one takes domain and codomain as separate arguments, as these are separate fields in the struct as well. I actually would prefer to store the homspace, and to make that one the most fundamental, but I didn't want to single-handedly decide this :)

Reply:
Ok, let's keep that for when we are completely changing and breaking the TensorMap structure.